Modules/Sources/Yosemite/PointOfSale/Items/PointOfSaleLocalBarcodeScanService.swift

```diff
@@ -0,0 +1,141 @@
+import Foundation
+import protocol Storage.GRDBManagerProtocol
+import class WooFoundation.CurrencySettings
+
+/// Service for handling barcode scanning using local GRDB catalog
+public final class PointOfSaleLocalBarcodeScanService: PointOfSaleBarcodeScanServiceProtocol {
+    private let grdbManager: GRDBManagerProtocol
+    private let siteID: Int64
+    private let itemMapper: PointOfSaleItemMapperProtocol
+
+    public init(siteID: Int64,
+                grdbManager: GRDBManagerProtocol,
+                currencySettings: CurrencySettings,
+                itemMapper: PointOfSaleItemMapperProtocol? = nil) {
+        self.siteID = siteID
+        self.grdbManager = grdbManager
+        self.itemMapper = itemMapper ?? PointOfSaleItemMapper(currencySettings: currencySettings)
+    }
+
+    /// Looks up a POSItem using a barcode scan string from the local GRDB catalog
+    /// - Parameter barcode: The barcode string from a scan (global unique identifier)
+    /// - Returns: A POSItem if found, or throws an error
+    public func getItem(barcode: String) async throws(PointOfSaleBarcodeScanError) -> POSItem {
+        do {
+            if let product = try searchProductByGlobalUniqueID(barcode) {
+                return try convertProductToItem(product, scannedCode: barcode)
+            }
+
+            if let variationAndParent = try searchVariationByGlobalUniqueID(barcode) {
+                return try await convertVariationToItem(variationAndParent.variation, parentProduct: variationAndParent.parentProduct, scannedCode: barcode)
+            }
+
+            throw PointOfSaleBarcodeScanError.notFound(scannedCode: barcode)
+        } catch let error as PointOfSaleBarcodeScanError {
+            throw error
+        } catch {
+            throw PointOfSaleBarcodeScanError.loadingError(scannedCode: barcode, underlyingError: error)
+        }
+    }
+
+    // MARK: - Product Search
+
+    private func searchProductByGlobalUniqueID(_ globalUniqueID: String) throws -> PersistedProduct? {
+        try grdbManager.databaseConnection.read { db in
+            try PersistedProduct.posProductByGlobalUniqueID(siteID: siteID, globalUniqueID: globalUniqueID).fetchOne(db)
+        }
+    }
+
+    // MARK: - Variation Search
+
+    private func searchVariationByGlobalUniqueID(_ globalUniqueID: String) throws -> (variation: PersistedProductVariation, parentProduct: PersistedProduct)? {
+        try grdbManager.databaseConnection.read { db in
+            guard let variation = try PersistedProductVariation.posVariationByGlobalUniqueID(siteID: siteID, globalUniqueID: globalUniqueID).fetchOne(db) else {
+                return nil
+            }
+            // Fetch parent product using the relationship
+            guard let parentProduct = try variation.request(for: PersistedProductVariation.parentProduct).fetchOne(db) else {
+                return nil
```

Comment:
It's a minor detail, but in the remote implementation, we throw `PointOfSaleBarcodeScanError.noParentProductForVariation` in case the parent product is not found

Reply:
As far as I can tell, it's not actually possible due to integrity constraints. If the parent is deleted, or not added in the first place, the variation would be deleted as well. I couldn't even make a test (involving the db) that the error is thrown... but I've added one that documents it shouldn't happen. b5f0a1f